src/motion_input_shaping/zero_vibration_stream_generator.py

```diff
@@ -0,0 +1,361 @@
+"""
+Contains the ZeroVibrationShaper class for re-use in other code.
+
+Run the file directly to test the class functionality.
+"""
+
+# pylint: disable=invalid-name
+# Allow short variable names
+
+import math
+from enum import Enum
+from dataclasses import dataclass
+import numpy as np
+
+
+class ShaperType(Enum):
+    """Enumeration for different input shaper types."""
+
+    ZV = 1
+    ZVD = 2
+    ZVDD = 3
+
+
+@dataclass
+class StreamSegment:
+    """A class that contains information for a single segment of the trajectory."""
+
+    position: float
+    speed_limit: float
+    accel: float
+    duration: float
+
+
+def trapezoidal_motion_generator(
+    distance: float, acceleration: float, deceleration: float, max_speed_limit: float
+) -> list[list[float]]:
+    """
+    Produce acceleration profile for basic trapezoidal motion.
+
+    Returns time and acceleration at points where acceleration changes
+    All acceleration changes are step changes for trapezoidal motion.
+
+    :param distance: The trajectory distance.
+    :param acceleration: The trajectory acceleration.
+    :param deceleration: The trajectory deceleration.
+    :param max_speed_limit: Maximum trajectory speed in the output motion.
+    """
+    direction = np.sign(distance)
+
+    acceleration_distance = max_speed_limit**2 / (2 * acceleration)
+    deceleration_distance = max_speed_limit**2 / (2 * deceleration)
+    max_speed_distance = abs(distance) - acceleration_distance - deceleration_distance
+
+    if max_speed_distance <= 0:
+        # Max speed is not reached.
+        max_speed = np.sqrt(abs(distance) / (1 / (2 * acceleration) + 1 / (2 * deceleration)))
+        acceleration_duration = max_speed / acceleration
+        deceleration_duration = max_speed / deceleration
+
+        acceleration_endtime = acceleration_duration
+        deceleration_endtime = acceleration_duration + deceleration_duration
+
+        return [
+            [0, acceleration_endtime, deceleration_endtime],
+            [acceleration * direction, deceleration * (-1 * direction), 0],
+        ]
+
+    # Trapezoidal Motion
+    max_speed = max_speed_limit
+
+    acceleration_duration = max_speed_limit / acceleration
+    max_speed_duration = max_speed_distance / max_speed_limit
+    deceleration_duration = max_speed_limit / deceleration
+
+    acceleration_endtime = acceleration_duration
+    max_speed_endtime = acceleration_endtime + max_speed_duration
+    deceleration_endtime = max_speed_endtime + deceleration_duration
+
+    return [
+        [0, acceleration_endtime, max_speed_endtime, deceleration_endtime],
+        [acceleration * direction, 0, deceleration * (-1 * direction), 0],
+    ]
+
+
+def calculate_acceleration_convolution(
+    impulse_times: list[float],
+    impulses: list[float],
+    unshaped_time: list[float],
+    unshaped_acceleration: list[float],
+) -> list[list[float]]:
```

Comment:
not clear what is returned.  Consider namedtuple or dataclass?

Reply:
Fixed with `AccelPoint` named tuple